src/main/kotlin/org/elm/workspace/ElmToolchain.kt

```diff
@@ -96,14 +98,31 @@ data class ElmToolchain(val binDirPath: Path) {
         return files.mapNotNull { Version.parseOrNull(it.name) }
     }
 
-    fun queryCompilerVersion(): Version? {
-        val elm = elmCompilerPath ?: return null
+    fun queryCompilerVersion(): Result<Version> {
+        val elm = elmCompilerPath ?: return Result.Err("Elm compiler not found")
+
+        // NodeJS tools like npm and nvm like to play games with wrapper scripts around native binaries
+        // such as the Elm compiler. So we will try to detect the wrapper and notify the user.
+        // See https://github.com/klazuka/intellij-elm/issues/252
+        try {
+            val firstChars = ByteArray(2)
+            elm.toFile().inputStream().read(firstChars, 0, 2)
+            if (firstChars.toString(Charsets.US_ASCII) == "#!") {
```

Comment:
@ajalt is there a more concise way to express this?

Also, do you know what the npm wrapper scripts would look like on Windows?

Reply:
If you don't mind reading a couple of extra bytes:

```kotlin
elm.toFile().bufferedReader().readLine()?.startsWith("#!")
```

The current code doesn't close the file. If that's necessary, you could do something like:

```kotlin
elm.toFile().bufferedReader().use { it.readLine() }?.startsWith("#!")
```

or

```kotlin
elm.toFile().useLines { it }.firstOrNull()?.startsWith("#!")
```

npm on windows creates two scripts in `%APPDATA%\Roaming\npm`.

One for cygwin:

```bash
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")

case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
esac

if [ -x "$basedir/node" ]; then
  "$basedir/node"  "$basedir/node_modules/elm/bin/elm" "$@"
  ret=$?
else 
  node  "$basedir/node_modules/elm/bin/elm" "$@"
  ret=$?
fi
exit $ret
```

and one for batch:

```batch
@IF EXIST "%~dp0\node.exe" (
  "%~dp0\node.exe"  "%~dp0\node_modules\elm\bin\elm" %*
) ELSE (
  @SETLOCAL
  @SET PATHEXT=%PATHEXT:;.JS;=;%
  node  "%~dp0\node_modules\elm\bin\elm" %*
)
```